go/libraries/doltcore/sqle/read_replica_database.go

```diff
@@ -230,3 +253,95 @@ func parseBranches(arg string) []string {
 	}
 	return branches
 }
+
+type res struct {
+	v   any
+	err error
+}
+
+type blocked struct {
+	f       func() (any, error)
+	waiters []chan res
+}
+
+func newLimiter() *limiter {
+	return &limiter{
+		blocked: make(map[string]*blocked),
+	}
+}
+
+// *limiter allows a caller to limit performing concurrent work for a given string key.
+type limiter struct {
+	mu      sync.Mutex
+	blocked map[string]*blocked
+}
+
+// |Run| invokes |f|, returning its result. It does not allow two |f|s
+// submitted with the same |s| to be running in concurrently.
+// Only one of the |f|s that arrives with the same |s| while another |f| with
+// that key is running will ultimately be run. The result of invoking that |f|
+// will be returned from the |Run| call to all blockers on that key.
+//
+// 1) A caller provides a string key, |s|, and an |f func() error| which will
+// perform the work when invoked.
+//
+// 2) If there is no outstanding call for the key, |f| is invoked and the
+// result is returned.
+//
+// 3) Otherwise, the caller blocks until the outstanding call is completed.
+// When the outstanding call completes, one of the blocked |f|s that was
+// provided for that key is run. The result of that invocation is returned to
+// all blocked callers.
+//
+// A caller's |Run| invocation can return early if the context is cancelled.
+// If the |f| captures a context, and that context is canceled, and the |f|
+// allows the error from that context cancelation to escape, then multiple
+// callers will see ContextCanceled / DeadlineExceeded, even if their contexts
+// are not canceled.
+//
+// This implementation is very naive and is not not optimized for high
+// contention on |l.blocked|/|l.mu|.
+func (l *limiter) Run(ctx context.Context, s string, f func() (any, error)) (any, error) {
+	l.mu.Lock()
+	if b, ok := l.blocked[s]; ok {
+		ch := make(chan res)
+		if b.f == nil {
+			b.f = f
+		}
+		b.waiters = append(b.waiters, ch)
+		l.mu.Unlock()
+		select {
+		case r := <-ch:
+			return r.v, r.err
+		case <-ctx.Done():
+			go func() { <-ch }()
+			return nil, ctx.Err()
+		}
+	} else {
+		l.blocked[s] = new(blocked)
+		l.mu.Unlock()
+	}
+
+	res, err := f()
+	l.finish(s)
+	return res, err
+}
+
+func (l *limiter) finish(s string) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	b := l.blocked[s]
+	if len(b.waiters) != 0 {
+		go func() {
+			r, err := b.f()
```

Comment:
so f` is always the second callback? even if the first routine is non-nil and there are > 2 routines?